packages/instruction-plans/src/transaction-planner-base.ts

```diff
@@ -0,0 +1,404 @@
+        appendTransactionMessageInstructions([instructionPlan.instruction], message);
+    const candidate = await selectAndMutateCandidate(context, context.parentCandidates, predicate);
+    if (candidate) {
+        return null;
+    }
+    const message = await createNewMessage(context, instructionPlan, predicate);
+    return { kind: 'single', message };
+}
+
+async function traverseMessagePacker(
+    instructionPlan: MessagePackerInstructionPlan,
+    context: TraverseContext,
+): Promise<MutableTransactionPlan | null> {
+    const messagePacker = instructionPlan.getMessagePacker();
+    const transactionPlans: SingleTransactionPlan[] = [];
+    const candidates = [...context.parentCandidates];
+
+    while (!messagePacker.done()) {
+        const candidate = await selectAndMutateCandidate(context, candidates, messagePacker.packMessageToCapacity);
+        if (!candidate) {
+            const message = await createNewMessage(context, instructionPlan, messagePacker.packMessageToCapacity);
+            const newPlan: MutableSingleTransactionPlan = { kind: 'single', message };
+            transactionPlans.push(newPlan);
+        }
+    }
+
+    if (transactionPlans.length === 1) {
+        return transactionPlans[0];
+    }
+    if (transactionPlans.length === 0) {
+        return null;
+    }
+    if (context.parent?.kind === 'parallel') {
+        return { kind: 'parallel', plans: transactionPlans };
+    }
+    return {
+        divisible: context.parent?.kind === 'sequential' ? context.parent.divisible : true,
+        kind: 'sequential',
+        plans: transactionPlans,
+    };
+}
+
+function getSequentialCandidate(latestPlan: MutableTransactionPlan): MutableSingleTransactionPlan | null {
+    if (latestPlan.kind === 'single') {
+        return latestPlan;
+    }
+    if (latestPlan.kind === 'sequential' && latestPlan.plans.length > 0) {
+        return getSequentialCandidate(latestPlan.plans[latestPlan.plans.length - 1]);
+    }
+    return null;
+}
+
+function getParallelCandidates(latestPlan: TransactionPlan): MutableSingleTransactionPlan[] {
+    return getAllSingleTransactionPlans(latestPlan);
+}
+
+async function selectAndMutateCandidate(
+    context: Pick<TraverseContext, 'abortSignal' | 'onTransactionMessageUpdated'>,
+    candidates: MutableSingleTransactionPlan[],
+    predicate: (
+        message: BaseTransactionMessage & TransactionMessageWithFeePayer,
+    ) => BaseTransactionMessage & TransactionMessageWithFeePayer,
+): Promise<MutableSingleTransactionPlan | null> {
+    for (const candidate of candidates) {
+        try {
+            const message = await getAbortablePromise(
+                Promise.resolve(
+                    context.onTransactionMessageUpdated(predicate(candidate.message), {
+                        abortSignal: context.abortSignal,
+                    }),
+                ),
+                context.abortSignal,
+            );
```

Comment:
Oh darn. I made this more difficult by not letting `getAbortablePromise()` take in a non-promise. Sorry.

You _could_ do the whole thenable thing here, but you'd have to think through the implications.
```suggestion
            const message = 'then' in context.onTransactionMessageUpdated
                ? await getAbortablePromise(
                       context.onTransactionMessageUpdated(predicate(candidate.message), {
                           abortSignal: context.abortSignal,
                       }),
                       context.abortSignal,
                   ),
                : context.onTransactionMessageUpdated(predicate(candidate.message), {
                      abortSignal: context.abortSignal,
                  });
```

Nah, what you've done here is better.